### PR TITLE
Add feature projection div cleaner

### LINKED ISSUE #4355: Initial B-field loading gives non divergence free field
When using an openPMD file to load an initial B-field, WarpX interpolates from the openPMD file's grid to the WarpX grid. This can lead to loss of the divergence free aspect of the B-field as demonstrated in the plot below where an FRC is loaded into WarpX:
image

Instead, the vector potential could be interpolated onto the WarpX grid (with staggering matching the E-field) and the B-field calculated as the curl of that field which ensures a divergence free B-field (up to machine precision):
image

Another alternative would be to perform some divergence cleaning of the B-field after interpolation.

### LINKED ISSUE #4356: Add warning when interpolating initial B-field
As a short-term remedy for #4355 a warning could be added that will alert the user of known issues with interpolating B-field values between non-matching grids.

### PR DESCRIPTION
**Pull request by clarkse**
Divergence cleaner class that computes the divergence at cell centers, then uses that as a source for a Poisson solve using MLMG that can be used to remove divergence components from the B-fields after initialization. This is useful for scrubbing interpolated B fields that are loaded into WarpX either through the openPMD loader or when using the pyWarpX fields interface.

### CODE REVIEW

**Comment by clarkse** [comment]
Fixes #4355 and #4356.

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (roelof-groenewald) [inline Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: Wouldn't the index stepping here cause issues in 1d/2d? Is there a reason not to use `UpwardD{x/y/z}` instead of `DownwardD{x/y/z}`?

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (roelof-groenewald) [inline Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp:202]: A cleaner approach could be to add a `ComputeDivB.cpp` file (analogous to https://github.com/ECP-WarpX/WarpX/blob/development/Source/FieldSolver/FiniteDifferenceSolver/ComputeDivE.cpp) that can be called from the `run_ProjectionDivCleaner` function to populate the source. What do you think?

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (clarkse) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: I had some issue with the prototype of UpwardD. I will see if I can adjust it so I don't need to explicitly index.

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (clarkse) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp:202]: I changed this to use the existing static ComputeDivB method from the WarpX class.

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (clarkse) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: This should now be moot as it has been removed in favor of using existing ComputeDivB.

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.H
Comment (roelof-groenewald) [inline Source/Initialization/DivCleaner/ProjectionDivCleaner.H]: Nice! We might want to start a coordinated effort to switch all WarpX header files to this approach instead of the old, lengthier 
```
#ifndef _MYHEADER_H_
#define _MYHEADER_H_
...
#endif
```

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (roelof-groenewald) [inline Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: Do you still plan on switching this to a warning in order to not break code that was previously running?

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (roelof-groenewald) [inline Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: This comment is outdated, right? Since the `lobc` and `hibc` are not necessarily Dirichlet.

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (roelof-groenewald) [inline Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: This comment also looks outdated. The function adds the gradient of the solution multifab to the simulation B-field, right?

File: Source/Initialization/WarpXInitData.cpp
Comment (roelof-groenewald) [inline Source/Initialization/WarpXInitData.cpp]: Typo:
```suggestion
        } else if (grid_type == GridType::Collocated) {
```

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.H
Comment (dpgrote) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.H]: The `#pragma once` looks like a nice feature but is not standard. This should be checked if it's appropriate here. Also, the header should follow the same pattern as other files, specifically "License: BSD-3-Clause-LBNL".

File: Source/Python/pyWarpX.cpp
Comment (roelof-groenewald) [inline Source/Python/pyWarpX.cpp]: Adding the Python binding here is a bit different from how similar things are done currently. The more standard approach would be that `run_ProjectionDivCleaner()` be a member function of `WarpX`, and the Python binding would then go in `Source/Python/WarpX.cpp`. Do you think it is worth making the change to follow that standard?

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (roelof-groenewald) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: Oh I see you added the warning in `WarpXInitData.cpp`.

**Review by roelof-groenewald — COMMENTED** [review]
Looks good! I think the namespace in which to put the calling function is the only remaining issue.

File: Source/Python/pyWarpX.cpp
Comment (clarkse) [reply Source/Python/pyWarpX.cpp]: Done

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.H
Comment (clarkse) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.H]: I think the pragma once is pretty widely supported. I switched over to include guards to be consistent with the rest of the header files.

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (roelof-groenewald) [inline Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]:
```suggestion
    // This function will compute -divB and store it in the source multifab
```

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (roelof-groenewald) [inline Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: Currently, there is no check that avoids the div B cleaner from being executed if the fft solver is used, right? In that case this error message would be confusing, but it would be good to simply not perform the div B cleaning when the fft solver is used, and print a message warning about possible B-field divergence (as is done for the other cases).

Comment (clarkse) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: Doesn't the PSATD solver have its own flavor of div cleaning that diffuses it away over time? I guess since the BCs could be incompatible it should not be executed if using the PSATD. I will add another check for that.

Comment (clarkse) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: I tightened this check further to keep it from being executed with incompatible solvers.

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (dpgrote) [inline Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: This line is only reached when `ProbLo(0) == 0`. What happens otherwise?

This whole section of code could be greatly simplified using a map.

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (dpgrote) [inline Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: Could this routine be used here? `InitializeStencilCoefficients`

File: Python/pywarpx/picmi.py
Comment (dpgrote) [inline Python/pywarpx/picmi.py]: This should be `self.do_divb_cleaning_external = kw.pop('warpx_do_divb_cleaning_external', None)` so that it has the same default as the compiled code and has the `warpx` prefix.

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (clarkse) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: I think it could be used here. I will attempt to do a refactor.

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (clarkse) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: I will take a look at it. We should probably also refactor the BC logic in the static solvers the same way. I grabbed similar logic from there and modified.  I think a separate PR would appropriate to clean up the BC logic in several spots.

Comment (roelof-groenewald) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: I think this probably needs something like
```
if (geom[0].ProbLo(0) == 0){
    lobc[0] = LinOpBCType::Neumann;
} else {
    // handle the r_min boundary explicitly
    if (WarpX::field_boundary_lo[0] == FieldBoundaryType::PEC) {
        lobc[0] = LinOpBCType::Dirichlet;
    }
    else if (WarpX::field_boundary_lo[0] == FieldBoundaryType::Neumann) {
        lobc[0] = LinOpBCType::Neumann;
    }
    else {
        WARPX_ALWAYS_ASSERT_WITH_MESSAGE(false,
            "Field boundary condition at the inner radius must be either PEC or neumann "
            "when using the Projection Divergence Cleaner"
        );
    }
}
```
at lines 123-124 with the rest of this block moved out of the `if` statement.

File: Source/WarpX.H
Comment (roelof-groenewald) [inline Source/WarpX.H]:
```suggestion
    //! Solve Poisson equation when loading an external magnetic field to clean divergence
```

File: Source/Initialization/WarpXInitData.cpp
Comment (roelof-groenewald) [inline Source/Initialization/WarpXInitData.cpp]: I might be missing something here but it seems like this is a bug in `development`, no? The external field should always be added to the fine patch fields, even if it is constant or parsed. So shouldn't this just be:
```suggestion
        if (m_p_ext_field_params->E_ext_grid_type != ExternalFieldType::default_zero) {
```

Assuming that is correct, I don't think the `load_from_python` type is needed anymore.

File: Examples/Tests/projection_divb_cleaner/inputs_rz
Comment (roelof-groenewald) [inline Examples/Tests/projection_divb_cleaner/inputs_rz:1]: I think for this and the `PICMI_inputs_3d` test in the same folder it would be better to modify the original tests instead of creating new ones. I say this because we know the existing tests are strictly less accurate than these new ones - not having divergence free B-fields are purely bad and therefore I don't see a good reason to keep both sets of tests.

**Review by roelof-groenewald — CHANGES_REQUESTED** [review]
A few minor comments.

I also think there is a bug in `development` for dealing with external grid fields. I will submit a PR to fix that which can then be updated in this PR as well.

File: Source/Initialization/WarpXInitData.cpp
Comment (clarkse) [reply Source/Initialization/WarpXInitData.cpp]: I believe the external fields are only allocated if the read_from_file, and now in this case the load_from_python flags are set for the external field grid type.

File: Examples/Tests/projection_divb_cleaner/inputs_rz
Comment (clarkse) [reply Examples/Tests/projection_divb_cleaner/inputs_rz:1]: So to be clear you want to merge the external field loader from file tests and have a separate one for the python loader?

File: Source/Initialization/WarpXInitData.cpp
Comment (roelof-groenewald) [reply Source/Initialization/WarpXInitData.cpp]: This can be updated once #5104 is merged.

File: Examples/Tests/projection_divb_cleaner/inputs_rz
Comment (roelof-groenewald) [reply Examples/Tests/projection_divb_cleaner/inputs_rz:1]: Yes, that sounds good to me.

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (clarkse) [reply Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp]: I updated this to use a std::map hash table.

File: Source/Initialization/WarpXInitData.cpp
Comment (clarkse) [reply Source/Initialization/WarpXInitData.cpp]: This has been merged from PR #5104

File: Examples/Tests/projection_divb_cleaner/inputs_rz
Comment (clarkse) [reply Examples/Tests/projection_divb_cleaner/inputs_rz:1]: @dpgrote When we talked earlier you thought that we should leave the default to off for the clean_divb_external flag. I think we should have a quick conversation between you, me, and @roelof-groenewald about it. I set it up to default to off. If we want to default it to on, then some of the other external field unit tests likely will need to have updated benchmarks. What behavior do we desire here?

Comment (clarkse) [reply Examples/Tests/projection_divb_cleaner/inputs_rz:1]: @roelof-groenewald Since the existing tests are for collocated grids, the cleaner with its default of being on will log a warning. The new CI tests added are on a Yee mesh, which will use the cleaner. I think covering all the options is best. I need to probably do some additional testing to make sure it is working properly with the collocated grids, but that could be another PR.

File: Python/pywarpx/picmi.py
Comment (roelof-groenewald) [inline Python/pywarpx/picmi.py]:
```suggestion
    in the beforeInitEsolve hook. This should operate identically to loading from
```

File: Python/pywarpx/picmi.py
Comment (roelof-groenewald) [inline Python/pywarpx/picmi.py]: I'm wondering about the default `True` here and in `LoadInitialField`. Usually we keep these default to `None` so that no parameter is passed to WarpX and therefore the C++ defaults are used (instead of needing to work at maintaining the same default settings in picmi and in C++ separately). Is it needed to have this default to `True`?

File: Python/pywarpx/picmi.py
Comment (roelof-groenewald) [inline Python/pywarpx/picmi.py]:
```suggestion
            pywarpx.warpx.do_divb_cleaning_external = self.do_divb_cleaning_external
```

File: Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp
Comment (roelof-groenewald) [inline Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp:160]: This is a nice and clean solution!

File: Python/pywarpx/picmi.py
Comment (roelof-groenewald) [inline Python/pywarpx/picmi.py]: It seems fine to set this to `None` by default since the C++ code will default it back to `True`.
```suggestion
        self.do_divb_cleaning_external = kw.pop('warpx_do_divb_cleaning_external', None)
```

File: Python/pywarpx/picmi.py
Comment (roelof-groenewald) [inline Python/pywarpx/picmi.py]: Same comment as below.
```suggestion
        self.do_divb_cleaning_external = kw.pop('warpx_do_divb_cleaning_external', None)
```

**Review by roelof-groenewald — COMMENTED** [review]
I think just a last couple minor defaulting things to sort out and we should be good to go!

File: Source/Initialization/WarpXInitData.cpp
Comment (roelof-groenewald) [inline Source/Initialization/WarpXInitData.cpp]: After we merge #5162, these lines could be moved to `WarpX::LoadExternalFields`.

File: Source/Python/WarpX.cpp
Comment (roelof-groenewald) [inline Source/Python/WarpX.cpp]:
```suggestion
            "Executes projection based divergence cleaner on loaded Bfield_fp_external."
```

File: Source/Initialization/WarpXInitData.cpp
Comment (clarkse) [reply Source/Initialization/WarpXInitData.cpp]: Sounds good. I will wait for that to be merged.

File: Regression/WarpX-tests.ini
Comment (roelof-groenewald) [inline Regression/WarpX-tests.ini]: These unused parameters were removed in #5043. 
```suggestion
```

File: Regression/WarpX-tests.ini
Comment (roelof-groenewald) [inline Regression/WarpX-tests.ini]:
```suggestion
```

File: Regression/WarpX-tests.ini
Comment (roelof-groenewald) [inline Regression/WarpX-tests.ini]:
```suggestion
```

File: Source/Initialization/WarpXInitData.cpp
Comment (clarkse) [reply Source/Initialization/WarpXInitData.cpp]: I moved this out of the check for checkpoint restart so it will run for new starts and restarts, however there were uninitialized values when trying to run from the LoadExternalFields routine, so I was unable to move to that location. I think having it run and cleaning it up prior to running initial solutions is a good spot.

**Review by roelof-groenewald — APPROVED** [review]
Thanks so much for adding this! Looks good to me.

**Pull request merged**
